### PR TITLE
changed authorization attribute 

### PR DESCRIPTION
**Pull request by Kunal-Jha**
ARUHA-2132
## Description
Added more parameters to authorization attribute 

## Review
- [ ] Tests
- [ ] Documentation
- [ ] CHANGELOG

### CODE REVIEW

File: src/main/java/org/zalando/nakadi/plugin/api/authz/AuthorizationService.java [hunk 1]
```diff
@@ -16,27 +16,30 @@
      * be accessible in the context, as done in org.springframework.security.core.context.SecurityContextHolder
      *
      * @param operation the operation (read, write, admin) to authorize
-     * @param resource the resource that the subject wants to perform an operation on
+     * @param resource  the resource that the subject wants to perform an operation on
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/plugin/api/authz/AuthorizationService.java]: space

File: src/main/java/org/zalando/nakadi/plugin/api/authz/AuthorizationService.java [hunk 2]
```diff
@@ -16,27 +16,30 @@
...
      * @return true if the subject, is authorized to perform the operation on the resource
      * @throws PluginException if an error occurred during execution
      */
     boolean isAuthorized(Operation operation, Resource resource) throws PluginException;
 
     /**
      * Check whether an attribute is valid.
-     *
+     * <p>
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/plugin/api/authz/AuthorizationService.java]: why is this needed?

File: src/main/java/org/zalando/nakadi/plugin/api/authz/AuthorizationService.java [hunk 3]
```diff
@@ -16,27 +16,30 @@
...
      * check that (a) 'username' is an accepted key, and (b) 'nakadi' is a username that exists and is active.
      *
      * @param attribute the attribute to validate
+     * @param resourceType the type of resource that the subject wants to perform an operation on
+     * @param operation the operation that the subject wants to perform on the resource
      * @return true if the attribute is valid
      * @throws PluginException if an error occurred during execution
      */
-    boolean isAuthorizationAttributeValid(AuthorizationAttribute attribute) throws PluginException;
+    boolean isAuthorizationAttributeValid(AuthorizationAttribute attribute, String resourceType, Operation operation)
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/plugin/api/authz/AuthorizationService.java]: 👍

File: src/main/java/org/zalando/nakadi/plugin/api/authz/AuthorizationService.java [hunk 4]
```diff
@@ -16,27 +16,30 @@
...
+            throws PluginException;
 
     /**
      * Filters a list of resources based on authorization rules.
-     *
+     * <p>
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/plugin/api/authz/AuthorizationService.java]: Is this needed?

**Comment by rcillo** [comment]
👍

File: src/main/java/org/zalando/nakadi/plugin/api/authz/EventAuthz.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package org.zalando.nakadi.plugin.api.authz;
+
+public interface EventAuthz
+{
+    public enum CompatibilityMode {
+        FORWARD, COMPATIBLE, NONE;
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/plugin/api/authz/EventAuthz.java]: If you have to replace all references to these enum values in Nakadi project it might be quite a big change affecting lots of files. Maybe you want to define a new method that "getCompatibilityMode" returning string and then implement it. It would cause a change only in the domain object EventTypeBase and not elsewhere.

File: src/main/java/org/zalando/nakadi/plugin/api/authz/EventAuthz.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    }
+    public enum CleanupPolicy {
+        DELETE, COMPACT
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/plugin/api/authz/EventAuthz.java]: The same as the comment above applies for this.

Comment (Kunal-Jha) [reply src/main/java/org/zalando/nakadi/plugin/api/authz/EventAuthz.java]: Resolved

File: src/main/java/org/zalando/nakadi/plugin/api/authz/EventAuthz.java [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package org.zalando.nakadi.plugin.api.authz;
+
+public interface EventAuthz
+{
+
+    String getCompatibilityMode();
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/plugin/api/authz/EventAuthz.java]: But I guess now that the names will conflict. I think you cannot have a method with the same name but two different return types.

Comment (Kunal-Jha) [reply src/main/java/org/zalando/nakadi/plugin/api/authz/EventAuthz.java]: Fixed

File: src/main/java/org/zalando/nakadi/plugin/api/authz/EventTypeAuthz.java [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package org.zalando.nakadi.plugin.api.authz;
+
+public interface EventTypeAuthz
+{
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/plugin/api/authz/EventTypeAuthz.java]: let's try to be more consistent with the formatting. Most `{` don't go in the new line. Also, regarding blank lines, we don't have strict rule for it, but usually we don't leave blank lines in between the last method definition and the closing `}`.

File: src/main/java/org/zalando/nakadi/plugin/api/authz/Resource.java [hunk 9]
```diff
@@ -13,4 +14,6 @@
 
     T get();
 
+    Map<String, List<AuthorizationAttribute>> getAuthorization();
+
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/plugin/api/authz/Resource.java]: I'm talking about this blank line.

File: src/main/java/org/zalando/nakadi/plugin/api/authz/AuthorizationService.java [hunk 10]
```diff
@@ -22,6 +23,18 @@
      */
     boolean isAuthorized(Operation operation, Resource resource) throws PluginException;
 
+    /**
+     * Check whether a Resource is valid.
+     *
+     * Example: Take an attribute with key 'username' and value 'nakadi'. A plugin implementing this method could
+     * check that (a) 'username' is an accepted key, and (b) 'nakadi' is a username that exists and is active.
```
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/plugin/api/authz/AuthorizationService.java]: I'm not sure id example description is correct, I think it describes a case for `boolean isAuthorizationAttributeValid(AuthorizationAttribute attribute)` method.

File: src/main/java/org/zalando/nakadi/plugin/api/authz/AuthorizationService.java (hunk 1 above)
Comment (rcillo) [reply src/main/java/org/zalando/nakadi/plugin/api/authz/AuthorizationService.java]: @Kunal-Jha comment was not addressed. Please, take a look.

File: src/main/java/org/zalando/nakadi/plugin/api/authz/AuthorizationService.java [hunk 11]
```diff
@@ -16,12 +19,27 @@
...
      * @return true if the subject, is authorized to perform the operation on the resource
      * @throws PluginException if an error occurred during execution
      */
     boolean isAuthorized(Operation operation, Resource resource) throws PluginException;
 
+    /**
+     * Checks whether all the permissions and actions for the resource are valid. It throws exception in case of problem
+     * and returns nothing otherwise.
+     *
+     * Example: Take a resource,for example, a subscription. A plugin implementing this method could
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/plugin/api/authz/AuthorizationService.java]: Please, add space after comma `resource,for` to `resource, for`.

File: src/main/java/org/zalando/nakadi/plugin/api/exceptions/OperationOnResourceNotPermitedException.java [hunk 12]
```diff
@@ -0,0 +1,8 @@
+package org.zalando.nakadi.plugin.api.exceptions;
+
+public class OperationOnResourceNotPermitedException extends RuntimeException {
+
+    public OperationOnResourceNotPermitedException(String message) {
+        super(message);
+    }
+}
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/plugin/api/exceptions/OperationOnResourceNotPermitedException.java]: Please add breakline at EOF.

**Comment by rcillo** [comment]
Looking good 👍 

Please, address minor comments so we can merge it.

File: src/main/java/org/zalando/nakadi/plugin/api/ApplicationService.java [hunk 13]
```diff
@@ -1,5 +1,7 @@
 package org.zalando.nakadi.plugin.api;
 
+import org.zalando.nakadi.plugin.api.exceptions.PluginException;
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/plugin/api/ApplicationService.java:3]: These interfaces import `PluginException` but no method mentions it explicitly. Isn't the idea to add it to the method signature?

File: src/main/java/org/zalando/nakadi/plugin/api/ApplicationServiceFactory.java [hunk 14]
```diff
@@ -1,5 +1,7 @@
 package org.zalando.nakadi.plugin.api;
 
+import org.zalando.nakadi.plugin.api.exceptions.PluginException;
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/plugin/api/ApplicationServiceFactory.java:3]: This one as well.

File: src/main/java/org/zalando/nakadi/plugin/api/ApplicationService.java (hunk 13 above)
Comment (Kunal-Jha) [reply src/main/java/org/zalando/nakadi/plugin/api/ApplicationService.java:3]: added to Signature

File: src/main/java/org/zalando/nakadi/plugin/api/ApplicationServiceFactory.java (hunk 14 above)
Comment (Kunal-Jha) [reply src/main/java/org/zalando/nakadi/plugin/api/ApplicationServiceFactory.java:3]: added to Signature

Comment (Kunal-Jha) [reply src/main/java/org/zalando/nakadi/plugin/api/ApplicationServiceFactory.java]: added to Signature

**Comment by rcillo** [comment]
👍

**Review by rcillo — APPROVED** [review]
